File: 1-week/set1/cash.c
#include <cs50.h>
#include <stdio.h>

int get_cents(void);
int calculate_quarters(int cents);
int calculate_dimes(int cents);
int calculate_nickels(int cents);
int calculate_pennies(int cents);
static int handle_coins(int cents, int base);

int main(void)
{
    // Ask how many cents the customer is owed
    int cents = get_cents();

    // Calculate the number of quarters to give the customer
    int quarters = calculate_quarters(cents);
    cents = cents - quarters * 25;

    // Calculate the number of dimes to give the customer
    int dimes = calculate_dimes(cents);
    cents = cents - dimes * 10;

    // Calculate the number of nickels to give the customer
    int nickels = calculate_nickels(cents);
    cents = cents - nickels * 5;

    // Calculate the number of pennies to give the customer
    int pennies = calculate_pennies(cents);
    cents = cents - pennies * 1;

    // Sum coins
    int coins = quarters + dimes + nickels + pennies;

    // Print total number of coins to give the customer
    printf("%i\n", coins);
}

int get_cents(void)
{
    int cents;
    do
    {
        cents = get_int("Change owed: ");
    }
    while (cents < 0);

    return cents;
}
/**
 * @brief (25C)
 *
 * @param cents
 * @return int
 */
int calculate_quarters(int cents)
{
    return handle_coins(cents, 25);
}

/**
 * @brief (10C)
 * @param cents
 * @return int
 */
int calculate_dimes(int cents)
{
    return handle_coins(cents, 10);
}

/**
 * @brief (5C)
 * @param cents
 * @return int
 */
int calculate_nickels(int cents)
{
    return handle_coins(cents, 5);
}

/**
 * @brief (1C)
 * @param cents
 * @return int
 */
int calculate_pennies(int cents)
{
    return handle_coins(cents, 1);
}

/**
 * @brief Handle the coins to return. Will calculate verify the base passed
 * @param cents value of cents
 * @param base Base to handle the calc
 * @return int Number of coins
 */
static int handle_coins(int cents, int base)
{
    int i = 0;
    while (cents > 0 && (cents / base) > 0)
    {
        cents = cents - base;
        i++;
    }
    return i;
}

File: 6-week/problem/sentimental-cash/cash.py
from cs50 import get_float


def main():
    cents = get_cents() * 100

    quarters = calc_quarters(cents)
    cents = cents - quarters * 25

    dimes = calc_dimes(cents)
    cents = cents - dimes * 10

    nickels = calc_nickels(cents)
    cents = cents - nickels * 5

    pennies = calc_pennies(cents)
    cents = cents - pennies * 1

    coins = quarters + dimes + nickels + pennies

    print(coins)


def get_cents():
    while True:
        cents = get_float("Change owed: ")
        if cents > 0:
            return cents


def calc_quarters(cents):
    return handle_coins(cents, 25)


def calc_dimes(cents):
    return handle_coins(cents, 10)


def calc_nickels(cents):
    return handle_coins(cents, 5)


def calc_pennies(cents):
    return handle_coins(cents, 1)


def handle_coins(cents, base):
    i = 0
    while (cents > 0 and int(cents / base) > 0):
        cents = cents - base
        i += 1
    return i


if __name__ == "__main__":
    main()
